pkg/controller/elasticsearch/driver/driver.go

```diff
@@ -155,14 +155,19 @@ func (d *defaultDriver) Reconcile() *reconciler.Results {
 
 	warnUnsupportedDistro(resourcesState.AllPods, d.ReconcileState.Recorder)
 
+	observerESClient, err := d.newElasticsearchClient(
+		resourcesState,
+		internalUsers.ControllerUser,
+		*min,
```

Comment:
What happens when we use v6 client to talk to v7 nodes? This will happen during upgrades, since, IIUC, we pick min version of all pods.

Reply:
> What happens when we use v6 client to talk to v7 nodes?

For most API calls, it will go through fine, but technically speaking, it is an unsupported scenario. Elasticsearch might add support for a request-specific version of the API to use (eg. in the header, as part of content negotiation), so that's something which might help.

Reply:
We already have a logic of switching from the generic cluster service URL to a particular Pod URL: https://github.com/elastic/cloud-on-k8s/blob/38ff3d3f4b4ff63f2db10afb6cd816fb4ebc8056/pkg/controller/elasticsearch/services/services.go#L107-L115
We needed it to handle the http/https migration. I think we could use something similar to handle version upgrades. If there are pods in 2 different versions, build a client for the lower version and use it to talk to that pods directly. So far that's not really a big problem though (when it is, we'll retry at next reconciliation anyway).